### PR TITLE
refactor-mesh-lookup - Use EnTT resource management for game assets

### PR DESCRIPTION
**Pull request by raffclar**

### CODE REVIEW

File: src/3D/L3DMesh.cpp
Comment (raffclar) [inline src/3D/L3DMesh.cpp:139]: Remove space above.

File: src/3D/MeshPack.cpp
Comment (raffclar) [inline src/3D/MeshPack.cpp:57]: Remove MeshPack (?).

File: src/3D/MeshPack.cpp
Comment (raffclar) [inline src/3D/MeshPack.cpp:62]: Remove MeshPack (?).

File: src/ECS/Systems/RenderingSystem.cpp
Comment (raffclar) [inline src/ECS/Systems/RenderingSystem.cpp:106]: Broken indent

File: src/Game.cpp
Comment (raffclar) [inline src/Game.cpp:340]: Remove this. Literals aren't being used.

File: src/Locator.h
Comment (raffclar) [inline src/Locator.h:29]: Not needed

File: src/Locator.h
Comment (raffclar) [inline src/Locator.h:30]: Not needed

File: src/Locator.h
Comment (raffclar) [inline src/Locator.h:55]: No EoL

File: src/Resources/ResourceManager.h
Comment (raffclar) [inline src/Resources/ResourceManager.h:71]: No EoL

File: src/3D/MeshPack.cpp
Comment (raffclar) [reply src/3D/MeshPack.cpp:62]: Removed.

**Comment by raffclar** [comment]
Fixing the CI failure. Looks like a change was made to fmtlib.

File: src/Resources/Loaders.h
Comment (bwrsandman) [inline src/Resources/Loaders.h:32]: Not of fan of using shared pointers. The reason for this is that it makes ownership impossible. The resource just stays around as long as there's a reference.
Isn't the point of using resource management that entt owns the resource?

File: src/Resources/Loaders.h
Comment (bwrsandman) [inline src/Resources/Loaders.h:55]: Could you add a loader for `.raw` textures from `filesystem.TexturePath()`

File: src/Resources/Resources.h
Comment (bwrsandman) [inline src/Resources/Resources.h:42]: EOL

File: src/Resources/Loaders.h
Comment (raffclar) [reply src/Resources/Loaders.h:32]: entt resource management relies on shared ptrs
https://github.com/skypjack/entt/blob/master/docs/md/resource.md

Comment (bwrsandman) [reply src/Resources/Loaders.h:32]: Then let's make our accesses with weak_ptr

Comment (raffclar) [reply src/Resources/Loaders.h]: Changed to use entt::resource_handle as the return type. It's basically an alias for shared_ptr but required by the interface spec.

File: src/Resources/Loaders.h
Comment (raffclar) [reply src/Resources/Loaders.h:55]: Yeah I can do that. Should I do this in a new PR? This is getting big.

**Comment by raffclar** [comment]
[7](https://github.com/openblack/openblack/pull/341/commits/7debda991a40313428679a79a38a13d6e75df1c1) will be dropped once approved

File: src/3D/AllMeshes.cpp
Comment (bwrsandman) [inline src/3D/AllMeshes.cpp:15]: In general +1 for moving this out of the header.
Could you make this a `constexpr` or `constinit` to make it threadsafe
Second, if there is any indication that this array can come from a resource file from the game, please make a note of it and write your code in function of the fact that in the future we want to load those strings from data files

File: src/ECS/Archetypes/FeatureArchetype.cpp
Comment (bwrsandman) [inline src/ECS/Archetypes/FeatureArchetype.cpp:43]: You can remove this line

File: src/ECS/Archetypes/Utils.cpp
Comment (bwrsandman) [inline src/ECS/Archetypes/Utils.cpp:29]: If they can be const please mark them as const by default

File: src/Game.cpp
Comment (bwrsandman) [inline src/Game.cpp:398]: Getting some warnings here about the type of `i = 0` being int instead of size_t

File: src/Game.cpp
Comment (bwrsandman) [inline src/Game.cpp:413]: here too

File: src/Game.cpp
Comment (bwrsandman) [inline src/Game.cpp:400]: Is `auto m` here to prevent a warning about disposing `m`?

File: src/Resources/Loaders.h
Comment (bwrsandman) [reply src/Resources/Loaders.h:55]: Yes, could you make an issue for that?

File: src/Resources/Loaders.h
Comment (bwrsandman) [inline src/Resources/Loaders.h:43]: Are you sure you want `nodiscard` here?

File: src/Resources/MeshId.h
Comment (bwrsandman) [inline src/Resources/MeshId.h:20]: Warnings about this function

**Review by bwrsandman — CHANGES_REQUESTED** [review]
Some changes needed but over all I would say it's very well done.
Great work!

File: src/Resources/Loaders.h
Comment (raffclar) [reply src/Resources/Loaders.h:55]: Added #369

File: src/3D/AllMeshes.cpp
Comment (raffclar) [reply src/3D/AllMeshes.cpp:15]: constexpr std::string support and constinit are both added in c++20

File: src/Game.cpp
Comment (raffclar) [reply src/Game.cpp:400]: Nicely spotted. Just debugging leftover

File: src/Resources/Loaders.h
Comment (raffclar) [reply src/Resources/Loaders.h:43]: Clang tidy tells me bad things if I remove them. Generally shouldn't be an issue as EnTT will be the caller of these `load` functions.

File: src/Resources/MeshId.h
Comment (raffclar) [reply src/Resources/MeshId.h:20]: image

Comment (raffclar) [reply src/Resources/MeshId.h:20]: So it's used but yeah this warning exists:
 Warning: /home/runner/work/openblack/openblack/src/Resources/MeshId.h:20:22: warning: unused function 'MeshIdToResourceId' [-Wunused-function]
  static entt::id_type MeshIdToResourceId(openblack::MeshId id)

Comment (bwrsandman) [reply src/Resources/MeshId.h:20]: Try putting it in a namespace see if that helps

File: src/ECS/Archetypes/FeatureArchetype.cpp
Comment (raffclar) [reply src/ECS/Archetypes/FeatureArchetype.cpp:43]: What do you mean?

Comment (raffclar) [reply src/ECS/Archetypes/FeatureArchetype.cpp]: Removed `mesh` assignment

File: src/Resources/Loaders.h
Comment (raffclar) [reply src/Resources/Loaders.h]: Is this ok?

**Comment by bwrsandman** [comment]
Will approve after some of the header code is moved to implementation files.

**Comment by raffclar** [comment]
> Will approve after some of the header code is moved to implementation files.

Moved the loader code to its own unit but I couldn't do much with the `ResourceManager` class. It doesn't have any methods tha can be refactored.

**Review by bwrsandman — APPROVED** [review]

**Comment by raffclar** [comment]
Rebasing

File: src/Resources/Loaders.cpp
Comment (bwrsandman) [inline src/Resources/Loaders.cpp:12]: You can just have `using namespace openblack::resources`. There's nothing new being defined

Comment (raffclar) [reply src/Resources/Loaders.cpp]: Fixed

**Pull request merged**
